File: mobile/lib/utils/lesson_ai_mindmap.dart
import 'dart:math' as math;

import '../models/chat_source_scope.dart';
import '../models/json_utils.dart';
import '../models/lesson.dart';
import '../models/lesson_analysis.dart';
import '../models/mind_map.dart';

const _centerX = 700.0;
const _centerY = 400.0;
const _branchRadius = 200.0;

List<String> collectMindMapBranchesFromSources(
  Lesson lesson,
  ChatSourceScope scope,
  List<LessonAnalysisEntry> analyses,
) {
  final branches = <String>{};

  for (final id in scope.analysisIds) {
    for (final a in analyses) {
      if (a.id != id) continue;
      final c = a.content;
      for (final k in (c['keyElements'] as List<dynamic>? ?? [])) {
        final t = k.toString().trim();
        if (t.isNotEmpty) branches.add(t);
      }
      for (final g in (c['grammarTopics'] as List<dynamic>? ?? [])) {
        final t = g.toString().trim();
        if (t.isNotEmpty) branches.add(t);
      }
      for (final r in (c['relatedConcepts'] as List<dynamic>? ?? [])) {
        final t = r.toString().trim();
        if (t.isNotEmpty) branches.add(t);
      }
      for (final n in (c['studyNotes'] as List<dynamic>? ?? [])) {
        final t = n.toString().trim();
        if (t.isNotEmpty) branches.add(t.length > 80 ? t.substring(0, 80) : t);
      }
      for (final ex in (c['exercises'] as List<dynamic>? ?? [])) {
        if (ex is! Map) continue;
        final title = ex['title']?.toString().trim() ?? '';
        if (title.isNotEmpty) {
          branches.add('تمرين ${ex['number']}: $title');
        }
      }
      for (final v in (c['vocabulary'] as List<dynamic>? ?? [])) {
        if (v is! Map) continue;
        final term = v['term']?.toString().trim() ?? '';
        if (term.isNotEmpty) {
          branches.add('$term: ${v['meaning'] ?? ''}');
        }
      }
      break;
    }
  }

  for (final id in scope.imageIds) {
    for (final img in lesson.images) {
      if (img.id != id) continue;
      final ai = img.aiAnalysis;
      if (ai != null) {
        branches.addAll(ai.keyElements);
        branches.addAll(ai.relatedConcepts);
      }
      break;
    }
  }

  for (final id in scope.noteIds) {
    for (final n in lesson.lessonNotes) {
      if (n.id == id && n.title.trim().isNotEmpty) {
        branches.add(n.title.trim());
        break;
      }
    }
  }

  for (final id in scope.wordPageIds) {
    for (final p in lesson.wordPages) {
      if (p.id == id && p.title.trim().isNotEmpty) {
        branches.add(p.title.trim());
        break;
      }
    }
  }

  return branches.take(24).toList();
}

List<MindMapNode> buildMindMapNodesFromSources(
  String lessonTitle,
  Lesson lesson,
  ChatSourceScope scope,
  List<LessonAnalysisEntry> analyses,
) {
  final branches = collectMindMapBranchesFromSources(lesson, scope, analyses);
  String center = lessonTitle;
  for (final a in analyses) {
    if (scope.analysisIds.contains(a.id)) {
      center = a.title;
      break;
    }
  }
  return buildMindMapNodesFromTexts(center, branches);
}

List<MindMapNode> buildMindMapNodesFromTexts(String centerText, List<String> branches) {
  final trimmed = branches.map((s) => s.trim()).where((s) => s.isNotEmpty).toList();
  if (centerText.trim().isEmpty && trimmed.isEmpty) return [];

  final centerId = newId();
  final nodes = <MindMapNode>[
    MindMapNode(
      id: centerId,
      text: centerText.trim().isEmpty ? '—' : centerText.trim(),
      x: _centerX - 84,
      y: _centerY - 26,
      parentId: null,
      color: hexFromColor(kMindMapNodeColors[0]),
      role: MindMapNodeRole.main,
    ),
  ];

  for (var i = 0; i < trimmed.length; i++) {
    final angle = (i / math.max(trimmed.length, 1)) * math.pi * 2 - math.pi / 2;
    final x = _centerX + math.cos(angle) * _branchRadius;
    final y = _centerY + math.sin(angle) * (_branchRadius * 0.65);
    nodes.add(
      MindMapNode(
        id: newId(),
        text: trimmed[i],
        x: x - 62,
        y: y - 19,
        parentId: centerId,
        color: hexFromColor(kMindMapNodeColors[(i + 1) % kMindMapNodeColors.length]),
        role: MindMapNodeRole.branch,
      ),
    );
  }
  return nodes;
}

File: lib/lesson-ai-mindmap.ts
import { buildMindMapNodesFromTexts } from "@/lib/mind-map-import"
import { getLessonNotes } from "@/lib/lesson-notes"
import { notePreviewText } from "@/lib/lesson-note-content"
import type { Lesson, MindMapNode } from "@/types/lesson"
import type { ChatSourceScope, LessonAnalysisEntry } from "@/types/lesson-analysis"

export type AiMindMapPlan = {
  title?: string
  categories?: Array<{ title?: string; note?: string; items?: Array<{ text?: string; note?: string }> }>
}

export function buildMindMapNodesFromAiPlan(plan: AiMindMapPlan, fallbackTitle: string): MindMapNode[] {
  const centerId = crypto.randomUUID()
  const categories = (plan.categories ?? []).filter((category) => category.title?.trim()).slice(0, 6)
  if (!categories.length) return []

  const nodes: MindMapNode[] = [{
    id: centerId,
    text: plan.title?.trim() || fallbackTitle,
    x: 616,
    y: 374,
    parentId: null,
    color: "#fef08a",
    role: "main",
    note: "خريطة ذهنية مولدة من تحليل الدرس بالذكاء الاصطناعي.",
  }]

  categories.forEach((category, categoryIndex) => {
    const angle = (categoryIndex / categories.length) * Math.PI * 2 - Math.PI / 2
    const categoryId = crypto.randomUUID()
    const x = 700 + Math.cos(angle) * 250
    const y = 400 + Math.sin(angle) * 180
    nodes.push({
      id: categoryId,
      text: category.title!.trim(),
      x: x - 72,
      y: y - 20,
      parentId: centerId,
      color: ["#bbf7d0", "#bfdbfe", "#fbcfe8", "#fed7aa", "#ddd6fe", "#ccfbf1"][categoryIndex],
      role: "branch",
      note: category.note?.trim() || "",
    })

    ;(category.items ?? []).filter((item) => item.text?.trim()).slice(0, 6).forEach((item, itemIndex) => {
      const offset = (itemIndex - ((category.items?.length ?? 1) - 1) / 2) * 58
      nodes.push({
        id: crypto.randomUUID(),
        text: item.text!.trim(),
        x: x + Math.cos(angle) * 165 - 60,
        y: y + Math.sin(angle) * 105 + offset - 17,
        parentId: categoryId,
        color: "#ffffff",
        role: "branch",
        note: item.note?.trim() || "",
      })
    })
  })
  return nodes
}

export function collectMindMapBranchesFromSources(
  lesson: Lesson,
  scope: ChatSourceScope,
  analyses: LessonAnalysisEntry[]
): string[] {
  const branches = new Set<string>()

  for (const id of scope.analysisIds) {
    const a = analyses.find((x) => x.id === id)
    if (!a) continue
    for (const k of a.content.keyElements ?? []) if (k.trim()) branches.add(k.trim())
    for (const g of a.content.grammarTopics ?? []) if (g.trim()) branches.add(g.trim())
    for (const c of a.content.relatedConcepts ?? []) if (c.trim()) branches.add(c.trim())
    for (const n of a.content.studyNotes ?? []) if (n.trim()) branches.add(n.trim().slice(0, 80))
    for (const ex of a.content.exercises ?? []) {
      if (ex.title.trim()) branches.add(`تمرين ${ex.number}: ${ex.title}`)
    }
    for (const v of a.content.vocabulary ?? []) {
      if (v.term.trim()) branches.add(`${v.term}: ${v.meaning}`)
    }
  }

  for (const id of scope.imageIds) {
    const img = lesson.images.find((i) => i.id === id)
    if (img?.aiAnalysis) {
      for (const k of img.aiAnalysis.keyElements) branches.add(k)
      for (const c of img.aiAnalysis.relatedConcepts) branches.add(c)
    }
  }

  const notes = getLessonNotes(lesson)
  for (const id of scope.noteIds) {
    const n = notes.find((x) => x.id === id)
    if (n?.title.trim()) branches.add(n.title.trim())
  }

  for (const id of scope.wordPageIds) {
    const p = (lesson.wordPages ?? []).find((w) => w.id === id)
    if (p?.title.trim()) branches.add(p.title.trim())
  }

  return [...branches].slice(0, 24)
}

export function buildMindMapNodesFromSources(
  lessonTitle: string,
  lesson: Lesson,
  scope: ChatSourceScope,
  analyses: LessonAnalysisEntry[]
): MindMapNode[] {
  const branches = collectMindMapBranchesFromSources(lesson, scope, analyses)
  const center =
    analyses.find((a) => scope.analysisIds.includes(a.id))?.title ||
    lessonTitle ||
    "الدرس"
  return buildMindMapNodesFromTexts(center, branches)
}
